internal/charmstore/migrations_dump_test.go

```diff
@@ -0,0 +1,715 @@
+package charmstore
+
+import (
+	"archive/zip"
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"go/build"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	jujutesting "github.com/juju/testing"
+	"github.com/juju/utils/fs"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/macaroon-bakery.v1/bakery"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+	"gopkg.in/tomb.v2"
+	"gopkg.in/yaml.v2"
+
+	"gopkg.in/juju/charmstore.v5-unstable/config"
+	"gopkg.in/juju/charmstore.v5-unstable/internal/blobstore"
+	"gopkg.in/juju/charmstore.v5-unstable/internal/router"
+)
+
+// historicalDBName holds the name of the juju database
+// as hard-coded in previous versions of the charm store server.
+const historicalDBName = "juju"
+
+// dumpMigrationDatabase checks out and runs the charmstore version held
+// in each element of history in sequence, runs any associated updates,
+// and, if the version is not before earlierDeployedVersion, dumps
+// the database to a file.
+//
+// After dumpMigrationDatabase has been called, createDatabaseAtVersion
+// can be used to backtrack the database to any of the dumped versions.
+func dumpMigrationHistory(session *mgo.Session, earliestDeployedVersion string, history []versionSpec) error {
+	db := session.DB(historicalDBName)
+	vcsStatus, err := currentVCSStatus()
+	if err != nil {
+		return errgo.Mask(err)
+	}
+	dumping := false
+	for _, vc := range history {
+		logger.Infof("----------------- running version %v", vc.version)
+		if vc.version == earliestDeployedVersion {
+			dumping = true
+		}
+		if err := runMigrationVersion(db, vc); err != nil {
+			return errgo.Notef(err, "cannot run at version %s", vc.version)
+		}
+		if dumping {
+			filename := migrationDumpFileName(vc.version)
+			logger.Infof("dumping database to %s", filename)
+			if err := saveDBToFile(db, vcsStatus, filename); err != nil {
+				return errgo.Notef(err, "cannot save DB at version %v", vc.version)
+			}
+		}
+	}
+	if !dumping {
+		return errgo.Newf("no versions matched earliest deployed version %q; nothing dumped", earliestDeployedVersion)
+	}
+	return nil
+}
+
+// createDatabaseAtVersion changes the state of the
+// database to be as it is after running the database
+// migrations from all the versions up to and including
+// the given version, which must be a version in the
+// dumped history not before the earliest deployed version.
```

Comment:
Is all this necessary? can't we just say it loads a database from a previously dumped file?